### PR TITLE
Feature/enkf q2

### LINKED ISSUE #566: A quick enhancement for clipping negative Q in the analysis of EnKF for fv3-lam
In the current FV3-LAM EnKF,  negative values in the hydrometers wold be replaced with the minimum positive values available in the running of this program.   But this clipping is not applied to the moisture (sphum ). The negative values would cause issues for the following modeling/forecast if the model wouldn't use clipping on its' own, that is the case found by @JiliDong-NOAA    
Hence,  a quick adding of such clipping for sphum (q) in the fv3-lam EnKF is to be done.

### PR DESCRIPTION
**Pull request by TingLei-daprediction**
Resubmit PR for cleaning the unneeded item in the previous PR 
In correspondence to the EMC GSI Issue#566, this PR contains a quick adding of the clipping of negative values of sphum (q) in the analysis of FV3-LAM EnKF.
This part of codes are not tested in the current GSI regression tests, which, hence, are not run.
The current codes are verified using local FV3-LAM case. It is found the differences from this changes exist for sphum ( maximum values about 0.003 (units) and 0.3 K for T (sensible T). The latter is because the sphum would be used when the analysis TV is converted to T. All differences are on spontaneous points and values are reasonable as expected.
Hence, the code is regarded as verified.

Fixes #566

**DUE DATE for this PR is 6/15/2023.** If this PR is not merged into `develop` by this date, the PR will be closed and returned to the developer.

### CODE REVIEW

**Comment by ShunLiu-NOAA** [comment]
Chunhua, David and Jeff,

Could you please review the change of limiting the negative values of sphum (q) in EnKF analysis?

Thank you.

Shun

**Review by chunhuazhou — DISMISSED** [review]
Looks good to me.

**Comment by daviddowellNOAA** [comment]
In the GFS (GDAS), are negative values of sphum corrected in the code / scripts?  Would the change to the EnKF code proposed here affect the GFS (GDAS)?

**Comment by TingLei-daprediction** [comment]
@daviddowellNOAA  In EnKF for GFS,  the clipping of negative values for Q is already implemented ( in gridio_gfs.f90). 
The change in this PR would not affect GDAS.

**Comment by daviddowellNOAA** [comment]
Ting, thanks for proposing these code changes and testing them.  As I understand things, the background specific humidity is clipped (to ensure values > 0) before the computation of virtual temperature, and then the analysis specific humidity is clipped before being output to a file.  Clipping makes sense in both instances, and I also see some parallels to how things are done in gridio_gfs.f90.  Looking at the code gave me some questions about future work.  Meanwhile, I support the changes in the current PR.

**Comment by daviddowellNOAA** [comment]
The comments below are beyond the scope of the current PR.  I'm providing them here for reference in future PRs.

(1) The EnKF for the global system (gridio_gfs.f90) appears to do clipping of the input values, before the analysis.  Should we consider the same for regional FV3?  I can see arguments for and against.

(2) Our current RRFS EnKF analyses (with radar-reflectivity DA) can produce negative values of precipitating hydrometeors.  Note that the clipping of precipitating hydrometeors in gridio_fv3reg.f90 only happens if l_use_enkf_directZDA is true, but l_use_enkf_directZDA is actually false in our system.  In the future, we should discuss if clipping should always happen for precipitating hydrometeors, no matter what the value of l_use_enkf_directZDA is.

**Comment by TingLei-daprediction** [comment]
@daviddowellNOAA  Good points. I will leave 1 for future discussions but I think I can  do 2 for this PR, namely, do clipping without dbz analysis being done.

**Comment by TingLei-daprediction** [comment]
@daviddowellNOAA  it is done as you commented , namely , changing " if (l_use_enkf_directZDA .and. cliptracers) then " to "if (cliptracers)". 
@chunhuazhou  there is a new change --very simple while meaningful.  Your further reviewing is appreciated.

**Review by jswhit — APPROVED** [review]

**Comment by ShunLiu-NOAA** [comment]
Thank you Jeff.

Shun

**Comment by ShunLiu-NOAA** [comment]
@chunhuazhou Could you have a look at this new change when you have a chance?

**Review by chunhuazhou — APPROVED** [review]

**Comment by ShunLiu-NOAA** [comment]
Thank you Chunhua.
Ting, could you please make a regression test with your changes?

Shun

**Comment by TingLei-daprediction** [comment]
@ShunLiu-NOAA   I have just merged the current EMC develop branch to this PR  and re-ran the verification for the fv3-lam EnKF.  It was as described in the previous verification, because in my case, the hydrometeors are not updated and hence the Yes/No to apply clipping to them doesn't affect the results.   @daviddowellNOAA  would you consider to run an enkf using dbz and see if the clipping of hydrometeors works as we expect ?   Thanks.

**Comment by ShunLiu-NOAA** [comment]
> @ShunLiu-NOAA I have just merged the current EMC develop branch to this PR and re-ran the verification for the fv3-lam EnKF. It was as described in the previous verification, because in my case, the hydrometeors are not updated and hence the Yes/No to apply clipping to them doesn't affect the results. @daviddowellNOAA would you consider to run an enkf using dbz and see if the clipping of hydrometeors works as we expect ? Thanks.

@TingLei-daprediction Could you kindly conduct a regression test for sanity before we commit codes to develop branch? Also, enkf run directories are available in /lfs/h2/emc/stmp/emc.lam/rrfs/v0.5.2/2023051708, could you please make a test on WCOSS2? Thank you.

**Comment by TingLei-daprediction** [comment]
@ShunLiu-NOAA  Thanks. I will finish those tests as soon as possible.

**Comment by daviddowellNOAA** [comment]
@TingLei-daprediction I successfully ran a summer 2022 case with EnKF assimilation of conventional observations followed by EnKF assimilation of radar-reflectivity observations.  The analysis looks good, and no negative values were produced for sphum, liq_wat, rainwat, snowwat, and graupel.  Thanks for working on these code changes!

**Comment by TingLei-daprediction** [comment]
@daviddowellNOAA  Thank you so much!  Then I will just need to finish the regular regression tests. 
Really appreciate your confirmation of the changes working as we expected.

**Comment by TingLei-daprediction** [comment]
@ShunLiu-NOAA  The netcdf_fv3_regional  regress test failed for the difference in the cost results between hiproc and loproc for both update and control runs.  Namely, the update runs are always identical to the control runs for both hiproc and loproc, while the differences are from loproc vs hiproc. 
Further investigation are needed.

**Comment by ShunLiu-NOAA** [comment]
Ting,

Thanks for the update.

Shun

**Comment by ShunLiu-NOAA** [comment]
@TingLei-daprediction Could you please make a regression test since the issue of netcdf_fv3_regional had been fixed?

**Comment by TingLei-daprediction** [comment]
@ShunLiu-NOAA  I will let you when the regression tests are finished.

**Comment by TingLei-daprediction** [comment]
@ShunLiu-NOAA  After mering the develop branch of  EMC GSI last week, now, all regression tests passed except for "rtma" for 
```
The case has Failed the scalability test.
The slope for the update (18.648673 seconds per node) is less than that for the control (19.833904 seconds per node
```
So, i think this is not an issue.

**Review by ShunLiu-NOAA — DISMISSED** [review]

File: src/enkf/gridio_fv3reg.f90
Comment (RussTreadon-NOAA) [inline src/enkf/gridio_fv3reg.f90]: Nitpick:  Why do we repeatedly set `clip` to `tiny` of a r_single variable?   It would more efficient to set `clip` once per subroutine instead of repeatedly using `tiny` to set it to the same value.

Comment (TingLei-daprediction) [reply src/enkf/gridio_fv3reg.f90]: I think the developer did this because he/she wanted to set the minimum positive value according to the variables's precision which could be varied (though, currently, they are the same).  I followed this "policy"  while assuming the cost of multiple "tiny()" actions are tiny. 
It is ok for me to use only one clip of r_single ( maybe, we could comment out (while leave there) those "tiny () to help other developers in the future notice the original developers' intent).

Comment (RussTreadon-NOAA) [reply src/enkf/gridio_fv3reg.f90]: The code is version controlled so please remove unused code.  Removing unused code improves readability.  If needed, developers can retrieve the previous version of the code via git commands.

**Comment by TingLei-daprediction** [comment]
@RussTreadon-NOAA .  I am not sure if developers would go back to the history of the coding to find the original intent if without any hints.  But it is ok.  Just one remaining question: did you mean only removing those "repeated" codes added in my PR, or should all the current "tiny" call be reduced to one?

**Comment by ShunLiu-NOAA** [comment]
@TingLei-daprediction as Russ pointed out, [your ](https://github.com/TingLei-daprediction/GSI/tree/feature/enkf_q2) is 2 commits behind NOAA-EMC:develop. Could you sync your branch and make a regression test again?
Thank you.

**Comment by TingLei-daprediction** [comment]
@ShunLiu-NOAA  I will let you know when I finish the suggested changes and the regression tests with the merged version.  Thanks.

**Comment by RussTreadon-NOAA** [comment]
I recommend cleaning up all occurrences of `clip` in `gridio_fv3reg.f90`.   The `if-then` blocks can also be collapsed to a single line.   

Set
```
    clip = tiny(1.0_r_single)
```
at the top of subroutines `writegriddata` and `writegriddata_pnc` 

Compress all the `if-then` clip blocks to a single `if` statement.  That is, rewrite
```
              if ( cliptracers ) then ! set cliptracers to remove negative hydrometers
                clip = tiny(qworkvar3d(1,1,1))
                where (qworkvar3d < clip) qworkvar3d = clip
              end if
```
as
```
              if ( cliptracers ) where (qworkvar3d < clip) qworkvar3d = clip  ! remove negative hydrometeors
```

This is clean, compact, and readable code.

**Comment by TingLei-daprediction** [comment]
@RussTreadon-NOAA  Thanks for this clarification.  I will keep all of you updated after I finish them.

**Comment by TingLei-NOAA** [comment]
An update: 
@ShunLiu-NOAA   After merging the current develop branch of EMC GSI and revisited the clip treatment as @RussTreadon-NOAA  suggested (while I used clip=tiny(clip) rather than clip =tiny(1.0_r_single)),  the current version reproduced the results of the prior version in my locale test case used in the previous verification.  The GSI regression tests only failed for the scalability tests in the hwrf_nmm_d3 and rtma, which are thought not an essential issue.

File: src/enkf/gridio_fv3reg.f90
Comment (RussTreadon-NOAA) [inline src/enkf/gridio_fv3reg.f90]: Update comment to accurately reflect what's being done.  `clip` is not zero.  `clip` is a tiny positive floating point value.  As such, the code does more than simply remove negative hydrometeors.  The code enforces a lower limit value of `tiny(clip)`, a positive number, on hydrometeors. 

Also, align the comment with code.  For example, something like
```
              call read_fv3_restart_data3d(varstrname,fv3filename,file_id,qworkvar3d)
              ! Enforce lower positive definite bound on hydrometers
              if ( cliptracers ) where (qworkvar3d < clip) qworkvar3d = clip
              tvworkvar3d=tsenworkvar3d*(one+fv*qworkvar3d)
```

Comment (ShunLiu-NOAA) [reply src/enkf/gridio_fv3reg.f90]: @TingLei-daprediction Could you please update this PR? Is it possible to consider including the bug fix below in this PR 
and make a regression test again? 

in innovstats.f90, around line 218:
^
call printstats('  all dbz',sumdbz_nh,biasq_nh,sumdbz_spread_nh,sumdbz_oberr_nh,nobsdbz_nh,&
V
the biasq_nh should be biasdbz_nh.

**Comment by TingLei-NOAA** [comment]
@RussTreadon-NOAA @ShunLiu-NOAA  The changes had been added as you suggested.  All regular GSI regression tests passed. 
Thanks

File: src/enkf/gridio_fv3reg.f90
Comment (RussTreadon-NOAA) [inline src/enkf/gridio_fv3reg.f90]: Nitpick:  remove extra space before `cliptracers` in `if` statement.   That is.  instead of 

              workvar3d=workvar3d+workinc3d
              if (  cliptracers ) where (workvar3d < clip) workvar3d = clip
              call write_fv3_restart_data3d(varstrname,fv3filename,file_id,workvar3d)

have

              workvar3d=workvar3d+workinc3d
              if ( cliptracers ) where (workvar3d < clip) workvar3d = clip
              call write_fv3_restart_data3d(varstrname,fv3filename,file_id,workvar3d)

File: src/enkf/gridio_fv3reg.f90
Comment (RussTreadon-NOAA) [inline src/enkf/gridio_fv3reg.f90]: Same nitpick comment as above.   Remove extra blank space before `cliptracers`

File: src/enkf/gridio_fv3reg.f90
Comment (RussTreadon-NOAA) [inline src/enkf/gridio_fv3reg.f90]: Same nitpick comment as above.   Remove extra blank space before `cliptracers`

**Comment by RussTreadon-NOAA** [comment]
@TingLei-daprediction , two questions
- on which machine(s) did you run the standard suite of 9 ctests?
- since none of the ctests exercise the modified code, have you run a test to confirm that the modified code achieves the desired end?

**Comment by TingLei-NOAA** [comment]
@RussTreadon-NOAA, the last time  I did the specific regression test which went through the changed code (gridio_fv3reg.f90) is the one as described 
^
An update:
@ShunLiu-NOAA After merging the current develop branch of EMC GSI and revisited the clip treatment as @RussTreadon-NOAA suggested (while I used clip=tiny(clip) rather than clip =tiny(1.0_r_single)), the current version reproduced the results of the prior version in my locale test case used in the previous verification. The GSI regression tests only failed for the scalability tests in the hwrf_nmm_d3 and rtma, which are thought not an essential issue.
V
But the recent code changes are two parts, the first is the changes on comments as you suggested which I didn't see actual impacts on the results (and for those very few changes , I think it is very unlikely there would be un noticed typo introduced in.    and the other is for the innovation for dbz obs .  Currently ,I don't have a dbz case as a reference  to test that and, such a change has been  tested before , and, similarly, it is very unlikely for some unnoticed typo introduced with such a tiny change, especially with the tracking function with this PR.   So, I didn't do the specific tests for the last changes.  Please let me know what you think.  Thanks

**Comment by TingLei-NOAA** [comment]
@RussTreadon-NOAA  removing all extra space before "cliptracer"  as you suggested.  Thanks. 
The code compiled successfully but no regression tests is done for the reasons I talked about in the previous conversion.

**Comment by ShunLiu-NOAA** [comment]
@TingLei-daprediction and @RussTreadon-NOAA I can do a regression test on WCOSS2.

**Comment by TingLei-NOAA** [comment]
@ShunLiu-NOAA  Thanks a lot!

**Comment by RussTreadon-NOAA** [comment]
Thank you, @ShunLiu-NOAA , for running the ctests on WCOSS2.   As mentioned before, our current suite of ctests do not exercise the modified code.   If we will be exercising the enkf in regional mode in operations, we should add a regional enkf regression test.

@TingLei-daprediction , you did not answer my question.  On which machine did you run the ctests?

Also, a fresh clone of  `TingLei-daprediction:feature/enkf_q2` at [99c6286](https://github.com/TingLei-daprediction/GSI/commit/99c6286376de595798625306b05dca6e2d60be15) now shows a variety of spacing on the `if ` `cliptracers` lines in `src/enkf/gridio_fv3reg.f90`

Below is a listing of all `if` lines with `cliptracers` in `gridio_fv3reg.f90`.
```
Orion-login-1:/work2/noaa/da/rtreadon/git/gsi/pr568$ grep -n "cliptracer" src/enkf/gridio_fv3reg.f90 | grep if
857:              if (cliptracers ) where (qworkvar3d < clip) qworkvar3d = clip
872:                 if (cliptracers ) where (qworkvar3d < clip) qworkvar3d = clip
939:             if (cliptracers ) where (workvar3d < clip) workvar3d = clip
959:              if (cliptracers ) where (workvar3d < clip) workvar3d = clip
979:             if (cliptracers ) where (workvar3d < clip) workvar3d = clip
999:             if (cliptracers ) where (workvar3d < clip) workvar3d = clip
1019:             if (  cliptracers ) where (workvar3d < clip) workvar3d = clip
1038:             if (  cliptracers ) where (workvar3d < clip) workvar3d = clip
2162:                  if (cliptracers ) where (qworkvar3d < clip) qworkvar3d = clip
2176:                     if (cliptracers ) where (qworkvar3d < clip) qworkvar3d = clip
2272:                if (cliptracers ) where (workvar3d < clip) workvar3d = clip
2303:                 if (  cliptracers ) where (workvar3d < clip) workvar3d = clip
2334:                 if (cliptracers ) where (workvar3d < clip) workvar3d = clip
2365:                 if (cliptracers ) where (workvar3d < clip) workvar3d = clip
2396:                if (  cliptracers ) where (workvar3d < clip) workvar3d = clip
2426:                if (cliptracers ) where (workvar3d < clip) workvar3d = clip
```

There is a double space between `if (` and `cliptracers` on lines 1019, 1038, 2303, and 2396.  The is no spacing between `if (` and `cliptracers` on lines 857, 872, 939, 959, 979, 999, 2162, 2176, 2272, 2334, 2365, and 2426.

It looks like you misunderstood my request.  I only wanted the double space on lines 1019, 1038, 2303, and 2396 changed to a single space.  The other lines did not need to be changed.   We want to see `if ( cliptracers )`.   There is one space before `cliptracers` and one space after `cliptracers`.

**Comment by ShunLiu-NOAA** [comment]
Thank you Russ. I will coordinate with Ming to update the regional
regression test case. It might take a while to prepare regional ensemble
members.

Shun

**Comment by TingLei-NOAA** [comment]
@RussTreadon-NOAA  Please if the current changes do the formatting as you requested.  Thanks.

**Comment by TingLei-NOAA** [comment]
@ShunLiu-NOAA  I think  a repeated regional enkf using dbz ob (from your operational runs)  should work for the regression test purpose ( go through the dbz related change).

**Comment by ShunLiu-NOAA** [comment]
@RussTreadon-NOAA Do you think it is acceptable if I run a test case with regional ensembles from realtime parallel without adding a new regression test? At the same time, we will work on prepare regional enkf regression test case for future ctest.

**Comment by RussTreadon-NOAA** [comment]
@TingLei-daprediction , a fresh clone of 40a6fd07 shows the desired spacing on the `if ( cliptracers )` lines.

As an aside, the way you update your forked branch `TingLei-daprediction:feature/enkf_q2` is problematic.   I should be able to `git pull` in my working copy of `TingLei-daprediction:feature/enkf_q2` after you update it.  I can't.  I get conflicts.  I suspect that you are not doing the update correctly.   The end result, however is OK.

**Comment by RussTreadon-NOAA** [comment]
> @RussTreadon-NOAA Do you think it is acceptable if I run a test case with regional ensembles from realtime parallel without adding a new regression test? At the same time, we will work on prepare regional enkf regression test case for future ctest.

What you propose is fine.  My ctest comment was looking ahead to ensure that future GSI PRs do not break the regional enkf or alter regional enkf output in unexpected ways.   We will be running the enkf in the operational RRFS, right?   If true, we should have a ctest which tests this future operational functionality.

**Review by RussTreadon-NOAA — COMMENTED** [review]
Changes look good.

**Comment by TingLei-NOAA** [comment]
> @TingLei-daprediction , a fresh clone of [40a6fd0](https://github.com/NOAA-EMC/GSI/commit/40a6fd07f90bb41d8765d804cd84f93715748eac) shows the desired spacing on the `if ( cliptracers )` lines.
> 
> As an aside, the way you update your forked branch `TingLei-daprediction:feature/enkf_q2` is problematic. I should be able to `git pull` in my working copy of `TingLei-daprediction:feature/enkf_q2` after you update it. I can't. I get conflicts. I suspect that you are not doing the update correctly. The end result, however is OK.

@RussTreadon-NOAA  not sure why.  I used "git push --force " to surpress some complaints from git.  I will try to resolve those complaints before using " git push " without "--force option to see if that could avoid the "conflict" complaints on your side. 
Thanks.

**Comment by RussTreadon-NOAA** [comment]
@TingLei-daprediction , you should not use `--force`.   Let's not push any more changes to this PR unless necessary.

**Comment by TingLei-NOAA** [comment]
@RussTreadon-NOAA Ok. Thanks

**Review by RussTreadon-NOAA — APPROVED** [review]
Approve

**Comment by ShunLiu-NOAA** [comment]
@RussTreadon-NOAA Thank you for approving this. I will merge the change after I complete my regional EnKF test.

**Comment by RussTreadon-NOAA** [comment]
Jeff's and Chunhua's reviews and approvals along with yours are what really matter.  My input is secondary.   Thank you @ShunLiu-NOAA for running a regional enkf test.  

We need to get a regional enkf test into the suite of ctests.   Additionally, since HAFSv1 is being implemented next week we should consider dropping the d2 and d3 HWRF tests in lieu of a new HAFS test.

**Comment by RussTreadon-NOAA** [comment]
This PR is 11 days past its 6/15/2023 due date.  This PR will be closed and returned to the developer tomorrow, 6/27/2023 unless it is finalized and merged into `develop` before then.

**Comment by ShunLiu-NOAA** [comment]
@Russ Treadon - NOAA Federal ***@***.***>  TIng and I are
testing EnKF with reflectivity assimilation. However, Ting's test with a
single case works, I still have troubles testing these changes with
realtime parallel. We are working on this.

**Comment by RussTreadon-NOAA** [comment]
Thank you @ShunLiu-NOAA for the update.   Testing delays are understood.   `TingLei-daprediction:feature/enkf_q2` will need to be updated as other PRs enter `develop`.  ctests and others tests will need to be rerun with the new baseline to ensure acceptable performance of the features added in this PR as well as not altering existing behavior in unexpected or unexplained ways.

**Comment by ShunLiu-NOAA** [comment]
Sure. If there is other PR merged in develop. We will run regression test again. By the way, we had regression tests on Friday. We also had an issue opened for regional EnKF test.

**Comment by RussTreadon-NOAA** [comment]
Yes, I saw GSI issue #586.   We should remove obsolete ctests as we add new tests.  We want to keep the total number of ctests at a reasonable number.

Who will replace the soon to be obsolete HWRF d2 and d3 ctests with a HAFS based ctest?    Do we need to update the RTMA test?  When can we retire the netcdf_fv3_regional test?

**Comment by ShunLiu-NOAA** [comment]
HAFS DA is planning to merge the changes in HAFS v1 back to GSI repo. After
this, we will use HAFS ctest to replace HWRF ctest. Let me get more
information about RTMA test and update you later.

Shun

**Review by ShunLiu-NOAA — APPROVED** [review]

**Pull request merged**
